### ISSUE TITLE
Unable to use on Windows Terraform

### ISSUE THREAD

**Issue by MarkLFT** [xunleii/terraform-module-k3s#95]
I have created a simple install to build a 3 node cluster.    My settings are as follows:

```
module "k3s" {
  source = "xunleii/k3s/module"

  depends_on = [
    time_sleep.wait_30_seconds
  ]

  drain_timeout  = "30s"
  managed_fields = ["label"] 

  generate_ca_certificates = false  # tried with and without this, same result

  global_flags = [
    "--tls-san 192.168.88.150"
  ]

  servers = {
    # The node name will be automatically provided by
    # the module using the field name... any usage of
    # --node-name in additional_flags will be ignored
    k3s_server_0 = {
      ip = "192.168.88.150" // internal node IP
      connection = {
        host     = "192.168.88.150" // public node IP
        user     = "mark"
        password = var.vm_password
      }
      labels = { "node.kubernetes.io/type" = "master" }
    }
    k3s_server_1 = {
      ip = "192.168.88.151"
      connection = {
        host     = "192.168.88.151" // public node IP
        user     = "mark"
        password = var.vm_password
      }
      labels = { "node.kubernetes.io/type" = "master" }
    }
    k3s_server_2 = {
      ip = "192.168.88.152"
      connection = {
        host     = "192.168.88.152" // public node IP
        user     = "mark"
        password = var.vm_password
      }
      labels = { "node.kubernetes.io/type" = "master" }
    }
  }
  agents = {}
}
```

However, when I run it, Terraform can ssh into the three Ubuntu hosts, and appears to be running normally.  Unit after about one minute, I receive the following warning and errors.

```
Warning: Argument is deprecated
│
│   with module.k3s.tls_self_signed_cert.kubernetes_ca_certs,
│   on .terraform\modules\k3s\k3s_certificates.tf line 42, in resource "tls_self_signed_cert" "kubernetes_ca_certs":
│   42:   key_algorithm         = "ECDSA"
│
│ This is now ignored, as the key algorithm is inferred from the `private_key_pem`.
│
│ (and 2 more similar warnings elsewhere)
╵
╷
│ Warning: expected allowed_uses to be one of [any_extended cert_signing client_auth code_signing content_commitment crl_signing data_encipherment decipher_only digital_signature email_protection encipher_only ipsec_end_system ipsec_tunnel ipsec_user key_agreement key_encipherment microsoft_commercial_code_signing microsoft_kernel_code_signing microsoft_server_gated_crypto netscape_server_gated_crypto ocsp_signing server_auth timestamping], got critical so will ignored
│
│   with module.k3s.tls_self_signed_cert.kubernetes_ca_certs,
│   on .terraform\modules\k3s\k3s_certificates.tf line 44, in resource "tls_self_signed_cert" "kubernetes_ca_certs":
│   44:   allowed_uses          = ["critical", "digitalSignature", "keyEncipherment", "keyCertSign"]
│
╵
╷
│ Warning: expected allowed_uses to be one of [any_extended cert_signing client_auth code_signing content_commitment crl_signing data_encipherment decipher_only digital_signature email_protection encipher_only ipsec_end_system ipsec_tunnel ipsec_user key_agreement key_encipherment microsoft_commercial_code_signing microsoft_kernel_code_signing microsoft_server_gated_crypto netscape_server_gated_crypto ocsp_signing server_auth timestamping], got digitalSignature so will ignored
│
│   with module.k3s.tls_self_signed_cert.kubernetes_ca_certs,
│   on .terraform\modules\k3s\k3s_certificates.tf line 44, in resource "tls_self_signed_cert" "kubernetes_ca_certs":
│   44:   allowed_uses          = ["critical", "digitalSignature", "keyEncipherment", "keyCertSign"]
│
╵
╷
│ Warning: expected allowed_uses to be one of [any_extended cert_signing client_auth code_signing content_commitment crl_signing data_encipherment decipher_only digital_signature email_protection encipher_only ipsec_end_system ipsec_tunnel ipsec_user key_agreement key_encipherment microsoft_commercial_code_signing microsoft_kernel_code_signing microsoft_server_gated_crypto netscape_server_gated_crypto ocsp_signing server_auth timestamping], got keyEncipherment so will ignored
│
│   with module.k3s.tls_self_signed_cert.kubernetes_ca_certs,
│   on .terraform\modules\k3s\k3s_certificates.tf line 44, in resource "tls_self_signed_cert" "kubernetes_ca_certs":
│   44:   allowed_uses          = ["critical", "digitalSignature", "keyEncipherment", "keyCertSign"]
│
╵
╷
│ Warning: expected allowed_uses to be one of [any_extended cert_signing client_auth code_signing content_commitment crl_signing data_encipherment decipher_only digital_signature email_protection encipher_only ipsec_end_system ipsec_tunnel ipsec_user key_agreement key_encipherment microsoft_commercial_code_signing microsoft_kernel_code_signing microsoft_server_gated_crypto netscape_server_gated_crypto ocsp_signing server_auth timestamping], got keyCertSign so will ignored
│
│   with module.k3s.tls_self_signed_cert.kubernetes_ca_certs,
│   on .terraform\modules\k3s\k3s_certificates.tf line 44, in resource "tls_self_signed_cert" "kubernetes_ca_certs":
│   44:   allowed_uses          = ["critical", "digitalSignature", "keyEncipherment", "keyCertSign"]
│
╵
╷
│ Error: remote-exec provisioner error
│
│   with module.k3s.null_resource.servers_install["k3s_server_2"],
│   on .terraform\modules\k3s\server_nodes.tf line 206, in resource "null_resource" "servers_install":
│  206:   provisioner "remote-exec" {
│
│ error executing "/tmp/terraform_814904232.sh": wait: remote command exited without exit status or exit signal
╵
╷
│ Error: remote-exec provisioner error
│
│   with module.k3s.null_resource.servers_install["k3s_server_1"],
│   on .terraform\modules\k3s\server_nodes.tf line 206, in resource "null_resource" "servers_install":
│  206:   provisioner "remote-exec" {
│
│ error executing "/tmp/terraform_1578112108.sh": wait: remote command exited without exit status or exit signal
╵
╷
│ Error: remote-exec provisioner error
│
│   with module.k3s.null_resource.servers_install["k3s_server_0"],
│   on .terraform\modules\k3s\server_nodes.tf line 206, in resource "null_resource" "servers_install":
│  206:   provisioner "remote-exec" {
│
│ error executing "/tmp/terraform_294316203.sh": wait: remote command exited without exit status or exit signal
```

I have tried both with and without the generate_ca_certificates, i.e. without defaults to true, and I have also tried to include it and set it to false.  I get the same results.

Am I missing something?  Does this only work on Linux boxes?  Can it be used against Ubuntu servers?   Any help that be provided will be greatly appreciated.

**Comment by MarkLFT** [comment]
By editing the k3s_certificates.tf file, I managed to get rid of the warnings.  

But I still get the remote-exec provisioner error.  This error is thrown at exactly 1m0s elapsed.  Which makes me think it is an ssh timeout rather than a problem with the code.   

When I check the hosts, it is obvious the instal process had started, but not finished before the ssh timeout.   Based on my research, there is no way on Ubuntu 20.04 to extend the timeout, it is based on TCP traffic only.   So I am not sure how I would go about resolving this.  

Does anyone else have the same problem?  And if so, how do you resolve it or workaround it?

**Comment by xunleii** [comment]
Hi @MarkLFT, thanks for your issue.
About the warning, it will be fixed by #88 (next release).

For this `1m` timeout, it is a bit weird; normally the default Terraform connection `timeout` value is `5m` (following https://www.terraform.io/language/resources/provisioners/connection#argument-reference). If you add `timeout = "10m"` inside the connection block, will it change anything?

Also, can you check your SSH server logs to see if it is not your SSH server that is closing it?

Thanks

**Comment by MarkLFT** [comment]
Many thank for the response.

I set the timeout to 5 mins, didn't make a difference. But after reading the Terraform Connection documentation, it appears this timeout is how long to wait to establish a connection, not a keep-alive type timeout.

Will check the logs to see what they say.
